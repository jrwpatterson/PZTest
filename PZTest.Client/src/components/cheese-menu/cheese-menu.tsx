import React from 'react'
import { makeStyles } from '@material-ui/core/styles'
import GridList from '@material-ui/core/GridList'
import GridListTile from '@material-ui/core/GridListTile'
import GridListTileBar from '@material-ui/core/GridListTileBar'
import { useCheeseMenu } from '../../hooks/use-cheese-menu'
import ListSubheader from '@material-ui/core/ListSubheader'
import IconButton from '@material-ui/core/IconButton'
import InfoIcon from '@material-ui/icons/Info'
import { Link } from 'react-router-dom'

const useStyles = makeStyles(theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    overflow: 'hidden',
    backgroundColor: 'lightgrey',
  },
  gridList: {
    height: 700,
  },
  subHeader: {
    height: 50,
  },
}))

export const CheeseMenu = () => {
  const { cheeses } = useCheeseMenu()
  const classes = useStyles()

  if (!cheeses) {
    return null
  }

  return (
    <div className={classes.root} data-testid='cheese.menu.root'>
      <GridList className={classes.gridList}>
        <GridListTile style={{ height: 'auto' }} cols={3} key='SubHeader'>
          <ListSubheader component='div' data-testid='cheese.menu.header'>
            Cheeses
          </ListSubheader>
        </GridListTile>
        {cheeses?.map(cheese => (
          <GridListTile key={cheese.id}>
            <Link to={`/detail/${cheese.id}`}>
              <img src={cheese.pictureUrl} alt={cheese.name} />
              <GridListTileBar
                data-testid='cheese.menu.tile'
                title={cheese.name}
                subtitle={<span>Price/Kg ${cheese.pricePerKG}</span>}
                actionIcon={
                  <IconButton
                    style={{ color: cheese.cheeseColour }}
                    aria-label={`info about ${cheese.name}`}
                  >
                    <InfoIcon />
                  </IconButton>
                }
              />
            </Link>
          </GridListTile>
        ))}
      </GridList>
    </div>
  )
}
