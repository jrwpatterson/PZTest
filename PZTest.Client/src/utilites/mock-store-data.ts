import { RootReducer } from '../root-reducer'

export const defaultMockStore: RootReducer = {
  basketReducer: {
    basket: {
      lines: [],
    },
  },
  cheeseReducer: {
    cheeses: [
      {
        cheeseColour: 'purple',
        id: 'testid',
        name: 'testCheese',
        pictureUrl: 'testUrl',
        pricePerKG: 10,
      },
      {
        cheeseColour: 'pink',
        id: 'testid2',
        name: 'testPinkCheese',
        pictureUrl: 'testPinkUrl',
        pricePerKG: 15,
      },
    ],
  },
}
