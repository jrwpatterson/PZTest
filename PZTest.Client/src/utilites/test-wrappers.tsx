import React, { ReactType, ReactNode } from 'react'
import { Provider } from 'react-redux'
import configureStore, { MockStoreEnhanced } from 'redux-mock-store'
import { defaultMockStore } from './mock-store-data'
import { render } from '@testing-library/react'
import { renderHook, RenderHookOptions } from '@testing-library/react-hooks'
import { RootReducer } from '../root-reducer'
// import { createMount } from '@material-ui/core/test-utils'
import { Provider as FetchProvider } from 'use-http'
import { MemoryRouter } from 'react-router-dom'

const mockStore = configureStore([])

const createStore = (storeData: RootReducer) => {
  return mockStore(storeData)
}

export const createMockProvider = (
  storeData: RootReducer = defaultMockStore,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={createStore(storeData)}>{children}</Provider>
    </FetchProvider>
  )
}

export const createMockProviderWithStore = (
  fakeStore: MockStoreEnhanced<unknown, {}>,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={fakeStore}>{children}</Provider>
    </FetchProvider>
  )
}

export const renderReduxConnectedComponent = (
  children: ReactType | JSX.Element,
  storeOverride: RootReducer = defaultMockStore,
  initialEntries?: string[],
) => {
  const storeData = storeOverride || defaultMockStore
  // const mount = createMount()
  const store = createStore(storeData)
  const Wrapper = createMockProviderWithStore(store)
  const element = (
    <MemoryRouter initialEntries={initialEntries}>
      <Wrapper>{children}</Wrapper>
    </MemoryRouter>
  )
  const component = render(element)

  return {
    store,
    ...component,
  }
}

export const renderReduxConnectedHook = <T extends {}>(
  callback: (props: unknown) => T,
  storeData: RootReducer = defaultMockStore,
  options: RenderHookOptions<unknown> = {},
) => {
  const store = createStore(storeData)
  const wrapper = createMockProviderWithStore(store)
  const renderedHook = renderHook(callback, { ...options, wrapper })

  return {
    store,
    ...renderedHook,
  }
}
