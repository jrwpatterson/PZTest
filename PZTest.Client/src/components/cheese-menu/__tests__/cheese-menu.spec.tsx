import React from 'react'
import { renderReduxConnectedComponent } from '../../../utilites/test-wrappers'
import { CheeseMenu } from '../cheese-menu'

describe('cheese-menu', () => {
  it('should match its snapshot', () => {
    const { container } = renderReduxConnectedComponent(<CheeseMenu />)
    expect(container).toMatchSnapshot()
  })

  it('should load a menu if a menu exists', () => {
    const { getByTestId } = renderReduxConnectedComponent(<CheeseMenu />)
    expect(getByTestId('cheese.menu.root')).toBeDefined()
  })

  it('should load the correct number of rows', () => {
    const { getAllByTestId } = renderReduxConnectedComponent(<CheeseMenu />)
    expect(getAllByTestId('cheese.menu.tile').length).toBe(2)
  })
})
