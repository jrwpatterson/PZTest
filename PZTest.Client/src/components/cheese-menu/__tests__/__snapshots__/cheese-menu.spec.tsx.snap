// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cheese-menu should match its snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
    data-testid="cheese.menu.root"
  >
    <ul
      class="MuiGridList-root makeStyles-gridList-2"
      style="margin: -2px;"
    >
      <li
        class="MuiGridListTile-root"
        style="width: 150%; height: auto; padding: 2px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <div
            class="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
            data-testid="cheese.menu.header"
          >
            Cheeses
          </div>
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 50%; height: 184px; padding: 2px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <a
            href="/detail/testid"
          >
            <img
              alt="testCheese"
              src="testUrl"
            />
            <div
              class="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
              data-testid="cheese.menu.tile"
            >
              <div
                class="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
              >
                <div
                  class="MuiGridListTileBar-title"
                >
                  testCheese
                </div>
                <div
                  class="MuiGridListTileBar-subtitle"
                >
                  <span>
                    Price/Kg $
                    10
                  </span>
                </div>
              </div>
              <div
                class="MuiGridListTileBar-actionIcon"
              >
                <button
                  aria-label="info about testCheese"
                  class="MuiButtonBase-root MuiIconButton-root"
                  style="color: purple;"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </a>
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 50%; height: 184px; padding: 2px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <a
            href="/detail/testid2"
          >
            <img
              alt="testPinkCheese"
              src="testPinkUrl"
            />
            <div
              class="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
              data-testid="cheese.menu.tile"
            >
              <div
                class="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
              >
                <div
                  class="MuiGridListTileBar-title"
                >
                  testPinkCheese
                </div>
                <div
                  class="MuiGridListTileBar-subtitle"
                >
                  <span>
                    Price/Kg $
                    15
                  </span>
                </div>
              </div>
              <div
                class="MuiGridListTileBar-actionIcon"
              >
                <button
                  aria-label="info about testPinkCheese"
                  class="MuiButtonBase-root MuiIconButton-root"
                  style="color: pink;"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </a>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
